### DISCUSSION TITLE
Use two I2c Bus

### DISCUSSION THREAD

**Discussion by Marco1971Repo** [pschatzmann/arduino-audio-driver#7] [Q&A] [answered]
I want connect a lcd display to the audio kit so i thought to use wire1 to initalize codec  pins 32, 33.
E wire to pilot display on standard pin 21, 22...

Do you think is possible or it is a total wrong idea?

**Comment by pschatzmann** [comment]
No sure, but if you search in the AudioKit issues, people were usually struggling to get this to work

**Comment by Marco1971Repo** [comment]
I modified the library i use to pilot lcd to use wire1 and it work.
I think could be great make available pin 21, 22 to connect i2c peripheral in a easy way.
If i was good enough in programming i have tried to do it to propose apullup but is far away to my capability.
Thank again for you support and work.

**Reply by uutzinger** [reply]
Not sure if this helps:

On some ESPs you can make multiple Wire instances but they still use a single Wire instance and you need to set SDA, SCL and clock each time before communicating with a device attached to different pins:

```
TwoWire   myWire = TwoWire(0);

void switchI2C(TwoWire *myPort, int sdaPin, int sclPin, uint32_t i2cSpeed) {
  myPort->begin(sdaPin, sclPin);
  myPort->setClock(i2cSpeed);
}
```

**Reply by Marco1971Repo** [reply]
Is the reason why i was suggesting to avoid to use default Wire object for internal use...So that it can be used on the externel pin with all the lybrary the use default Wire object for i2c....

Is an idea but i'm not good enough in programming to understand if it is feasible and how it is complex
In my case i've modified hd44780 library to use Wire1 instead of Wire doing Wire1.begin(21, 22)...

**Reply by pschatzmann** [reply]
This might work for the ESP32 but not for other architectures. 
If you are not happy with the default assignment, you can still change it via the API or you can create your own board definition where you specify the Wire object

**Comment by Marco1971Repo** [comment]
I Phil need you help if you can.
Following this discussion, I tought (after you merged my pull request) to do:
```
  AudioKitEs8388V1.begin(cfg); 

//Wire1 to internal use
  AudioKitEs8388V1.getPins().addI2C(PinFunction::CODEC, 32, 33, -1, 100000U, Wire1);

  AudioKitEs8388V1.setVolume(iInitialVolume);
  KEY_1.attachClick(prevStation);

```
But it has absolutely no effect.
If i change this line of code in PinsAudioKitEs8388v1Class:
```
    // add i2c codec pins: scl, sda, port, frequency
    addI2C(PinFunction::CODEC, 32, 33);
```
to:
```
    // add i2c codec pins: scl, sda, port, frequency
    addI2C(PinFunction::CODEC, 32, 33, -1, 100000U, Wire1);
```
It work exactly as i thought. and everything work fine.

How is the correct wayin your opinion to pass the last  3 parameters to addI2C(...), so to use Wire1 internally instead of Wire, in your architecture?

**Comment by pschatzmann** [comment]
AudioKitEs8388V1.getPins().addI2C(PinFunction::CODEC, 32, 33, -1, 100000U, Wire1);
is adding a new line, but this is not what you want!
You want to just update the I2C definitin on the existing line!
```
auto i2c = getPins().getI2C(PinFunction::CODEC);
if (i2c) i2c.value().p_wire = &Wire1;
```

**Reply by Marco1971Repo** [reply]
It has no effect.
Or maybe i haven't understand your suggestion.
This i my code:
```
#include "Arduino.h"
#include "WiFi.h"
#include "Audio.h"
#include "AudioBoard.h"
#include "OneButton.h"

#include <hd44780.h>                       // main hd44780 header
#include <hd44780ioClass/hd44780_I2Cexp.h> // i2c expander i/o class header
hd44780_I2Cexp lcd; // declare lcd object: auto locate & auto config expander chip
// LCD geometry
const int LCD_COLS = 20;
const int LCD_ROWS = 4;
unsigned long iUltimaAccensioneDisplay = 0;
unsigned int iTimeoutDisplay = 5000;
bool bLcdFatalError = false;
//Audiokit i2s pin definition
#define I2S_DOUT      26 //35
#define I2S_BCLK      27
#define I2S_LRC       25
Audio audio;
OneButton KEY_1(36), KEY_2(13), KEY_3(19), KEY_4(23);

unsigned int iInitialVolume = 50;

void nextStation();
void prevStation();
void volumeDown();
void volumeUp();
void printOnLcd(int idx, const char *info = NULL);


const char* ssid_1     = "FASTWEB-FE2QBJ-24";         // The SSID (name) of the Wi-Fi network you want to connect to
//const char* ssid_1     = "iPhone di Marco";         // The SSID (name) of the Wi-Fi network you want to connect to
const char* password_1 = "RRNI6SEGN3";     // The password of the Wi-Fi network
#define IDX_LAST_STATIONS 9
static int i_stationIdx = -1;
const char *stationsName[] = {
  PROGMEM("Radio Deejay"),
  PROGMEM("Deejay 80"),
  PROGMEM("Deejay On The Road"),
  PROGMEM("Tropical Pizza"),
  PROGMEM("Virgin Radio"),
  PROGMEM("Mitology"), //Mytology
  PROGMEM("Controradio"), //Controradio  
  PROGMEM("RTL 102.5"),
  PROGMEM("Radio 105"),
  PROGMEM("Subasio")
};
const char *stationUrls[] = {
  PROGMEM("http://streamcdnb1-4c4b867c89244861ac216426883d1ad0.msvdn.net/radiodeejay/radiodeejay/play1.m3u8"),
  PROGMEM("http://streamcdnf25-4c4b867c89244861ac216426883d1ad0.msvdn.net/webradio/deejay80/live.m3u8"),
  PROGMEM("http://streamcdnm5-4c4b867c89244861ac216426883d1ad0.msvdn.net/webradio/deejayontheroad/live.m3u8"),
  PROGMEM("http://streamcdnm12-4c4b867c89244861ac216426883d1ad0.msvdn.net/webradio/deejaytropicalpizza/live.m3u8"),
  PROGMEM("http://icecast.unitedradio.it/Virgin.mp3"),
  PROGMEM("http://onair15.xdevel.com/proxy/radiocatsdogs2?mp=/;stream/;"), //Mytology
  PROGMEM("http://s4.yesstreaming.net:7199/stream"), //Controradio  
  PROGMEM("http://streamingv2.shoutcast.com/rtl-1025"),
  PROGMEM("http://icecast.unitedradio.it/Radio105.mp3"),
  PROGMEM("http://icy.unitedradio.it/Subasio.mp3")
};
void setup() {
    
    Serial.begin(115200);
    while (!Serial) {
    ;
    }
    LOGLEVEL_AUDIODRIVER = AudioDriverDebug;
    WiFi.disconnect();
    WiFi.mode(WIFI_STA);
    WiFi.begin(ssid_1, password_1);
    while (WiFi.status() != WL_CONNECTED)
    delay(1500);

    Serial.printf("Total heap: %d\n", ESP.getHeapSize());
    Serial.printf("Free heap: %d\n", ESP.getFreeHeap());
    Serial.printf("Total PSRAM: %d\n", ESP.getPsramSize());
    Serial.printf("Free PSRAM: %d-%d\n", ESP.getFreePsram(), UINT16_MAX);
    Serial.printf( "ESP32 Chip model:  %s Rev %d\n", ESP.getChipModel(), ESP.getChipRevision());
    Serial.printf( "SdkVersion:        %s\n",                            ESP.getSdkVersion());
   
    audio.setPinout(I2S_BCLK, I2S_LRC, I2S_DOUT, 0);
    audio.setVolume(21);
    CodecConfig cfg;
    cfg.input_device = audio_driver::ADC_INPUT_ALL;
    cfg.output_device = audio_driver::DAC_OUTPUT_ALL;
    cfg.i2s.bits = audio_driver::BIT_LENGTH_16BITS;
    cfg.i2s.rate = audio_driver::RATE_44K;
    cfg.i2s.fmt = audio_driver::I2S_NORMAL;
    //cfg.i2s.mode = audio_driver::MODE_SLAVE;

    AudioKitEs8388V1.begin(cfg); 
    auto i2c = AudioKitEs8388V1.getPins().getI2CPins(PinFunction::CODEC);
    if (i2c)
     i2c.value().p_wire = &Wire1;
    AudioKitEs8388V1.setVolume(iInitialVolume);
    KEY_1.attachClick(prevStation);
    KEY_2.attachClick(nextStation);
    KEY_3.attachClick(volumeDown);
    KEY_4.attachClick(volumeUp);

    
   
  	int status = lcd.begin(LCD_COLS, LCD_ROWS);
	if(status) // non zero status means it was unsuccesful
	{
        bLcdFatalError =true;
	} else
    {
        lcd.noBacklight(); //spengiamo lo lcd
        lcd.lineWrap();
    }
    nextStation();   
}
void loop()
{
  /*size_t psram_size = esp_spiram_get_size();
Serial.printf("PSRAM size: %d bytes\n", psram_size);*/
    audio.loop();
    KEY_1.tick();
    KEY_2.tick();
    KEY_3.tick();
    KEY_4.tick();
    if ((millis() - iUltimaAccensioneDisplay) > iTimeoutDisplay)
    {
        lcd.noBacklight();
        iUltimaAccensioneDisplay = millis();
    }
}

void volumeDown()
{
    if (AudioKitEs8388V1.getVolume() > 0)
        iInitialVolume -= 5;
    AudioKitEs8388V1.setVolume(iInitialVolume);
    printOnLcd(i_stationIdx);
}
void volumeUp()
{
    //audio.setTone(10, -10,-10);
    if (AudioKitEs8388V1.getVolume() < 100)
        iInitialVolume += 5;
    AudioKitEs8388V1.setVolume(iInitialVolume);
    printOnLcd(i_stationIdx);
}
void prevStation()
{
    i_stationIdx--;
    if (i_stationIdx < 0)
        i_stationIdx = IDX_LAST_STATIONS;
    if(!audio.connecttospeech(stationsName[i_stationIdx], "it")) // Google TTS
        audio.connecttohost(stationUrls[i_stationIdx]);
    lcd.clear();
    lcd.printf("%s", stationsName[i_stationIdx]);
    printOnLcd(i_stationIdx);
} 
void nextStation()
{
    if (i_stationIdx < IDX_LAST_STATIONS)
        i_stationIdx++;
    else
        i_stationIdx = 0;
    Serial.printf(">>>>>>>>>>>>>>>>>nextStation %d\n", i_stationIdx);
    if(!audio.connecttospeech(stationsName[i_stationIdx], "it")) // Google TTS
        audio.connecttohost(stationUrls[i_stationIdx]);
    printOnLcd(i_stationIdx);
} 

void printOnLcd(int idx, const char* info)
{
    if (!bLcdFatalError)
    {
        lcd.clear();
        lcd.setCursor(0, 0);
        lcd.printf("Sta.:%d", idx + 1);
        lcd.setCursor(0, 1);
        lcd.printf("%s", stationsName[idx]);
        lcd.setCursor(10, 0);
        lcd.printf("Vol.: %d%%", AudioKitEs8388V1.getVolume());
        iUltimaAccensioneDisplay = millis();
        lcd.backlight();
        if (info)
        {
            String sInfo = info;
            String sAppo = "";
            int iIndx = 0;
            sInfo.trim();
            //sInfo.replace('~', ' ');
            lcd.setCursor(0, 2);
            iIndx = sInfo.indexOf('~');
            if (iIndx > 0)
            {
                sAppo = sInfo.substring(0, iIndx);
                lcd.printf("%s", sAppo.c_str());
                lcd.setCursor(0, 3);
                sAppo = sInfo.substring(iIndx+1, iIndx + 20);
                iIndx = sAppo.indexOf('~');
                lcd.printf("%s", sAppo.substring(0, iIndx).c_str());
                return;
            }
            iIndx = sInfo.indexOf('-');
            if (iIndx > 0)
            {
                sAppo = sInfo.substring(0, iIndx);
                lcd.printf("%s", sAppo.c_str());
                lcd.setCursor(0, 3);
                sAppo = sInfo.substring(iIndx+1, iIndx + 20);
                sAppo.trim();
                lcd.printf("%s", sAppo.substring(0, 19).c_str());
                return;
            }else
                lcd.printf("%s", sInfo.substring(0, 39).c_str());
        }
    }
} 

void audio_showstreamtitle(const char* info)
{
    Serial.printf("showstreamtitle %s\n", info);
    printOnLcd(i_stationIdx, info);
}
void audio_icydescription(const char* info)
{
    Serial.printf("icydescription %s\n", info);
}
void audio_commercial(const char* info)
{
    Serial.printf("commercial %s\n", info);
}

void audio_eof_speech(const char*info)
{
    audio.connecttohost(stationUrls[i_stationIdx]);
}
```

**Comment by pschatzmann** [comment]
you need to do this before calling begin() otherwise it has no effect

**Reply by Marco1971Repo** [reply]
No...nothing change, log with lines moved before begin:
`␛[0;32mI (39) boot.esp32: SPI Mode       : DIO␛[0m
␛[0;32mI (43) boot.esp32: SPI Flash Size : 4MB␛[0m
␛[0;32mI (48) boot: Enabling RNG early entropy source...␛[0m
␛[0;32mI (53) boot: Partition Table:␛[0m
␛[0;32mI (57) boot: ## Label            Usage          Type ST Offset   Length␛[0m  
␛[0;32mI (64) boot:  0 nvs              WiFi data        01 02 00009000 00004000␛[0m
␛[0;32mI (72) boot:  1 phy_init         RF data          01 01 0000d000 00001000␛[0m
␛[0;32mI (79) boot:  2 factory          factory app      00 00 00010000 00300000␛[0m
␛[0;32mI (87) boot: End of partition table␛[0m
␛[0;32mI (91) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=46fcch (290764) map␛[0m
␛[0;32mI (200) esp_image: segment 1: paddr=00056ff4 vaddr=3ffb0000 size=03308h ( 13064) load␛[0m
␛[0;32mI (205) esp_image: segment 2: paddr=0005a304 vaddr=40080000 size=05d14h ( 23828) load␛[0m
␛[0;32mI (215) esp_image: segment 3: paddr=00060020 vaddr=400d0020 size=db1dch (897500) map␛[0m
␛[0;32mI (525) esp_image: segment 4: paddr=0013b204 vaddr=40085d14 size=12164h ( 74084) load␛[0m
␛[0;32mI (565) boot: Loaded app from partition at offset 0x10000␛[0m
␛[0;32mI (565) boot: Disabling RNG early entropy source...␛[0m
␛[0;32mI (577) cpu_start: Multicore app␛[0m
␛[0;32mI (577) psram: This chip is ESP32-D0WD␛[0m
␛[0;32mI (577) spiram: Found 64MBit SPI RAM device␛[0m
␛[0;32mI (579) spiram: SPI RAM mode: flash 80m sram 80m␛[0m
␛[0;32mI (584) spiram: PSRAM initialized, cache is in low/high (2-core) mode.␛[0m
␛[0;32mI (591) cpu_start: Pro cpu up.␛[0m
␛[0;32mI (595) cpu_start: Starting app cpu, entry point is 0x40083ba4␛[0m
␛[0;32mI (0) cpu_start: App cpu up.␛[0m
␛[0;32mI (1077) spiram: SPI SRAM memory test OK␛[0m
␛[0;32mI (1085) cpu_start: Pro cpu start user code␛[0m
␛[0;32mI (1085) cpu_start: cpu freq: 240000000␛[0m
␛[0;32mI (1085) cpu_start: Application information:␛[0m
␛[0;32mI (1088) cpu_start: Project name:     240219-195323-esp32dev␛[0m
␛[0;32mI (1094) cpu_start: App version:      28d3ef6-dirty␛[0m
␛[0;32mI (1100) cpu_start: Compile time:     Mar  5 2024 15:47:22␛[0m
␛[0;32mI (1106) cpu_start: ELF file SHA256:  021af9d9ba89a91d...␛[0m
␛[0;32mI (1112) cpu_start: ESP-IDF:          4.4.6␛[0m
␛[0;32mI (1117) cpu_start: Min chip rev:     v0.0␛[0m
␛[0;32mI (1122) cpu_start: Max chip rev:     v3.99 ␛[0m
␛[0;32mI (1127) cpu_start: Chip rev:         v3.0␛[0m
␛[0;32mI (1132) heap_init: Initializing. RAM available for dynamic allocation:␛[0m
␛[0;32mI (1139) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM␛[0m
␛[0;32mI (1145) heap_init: At 3FFB9380 len 00026C80 (155 KiB): DRAM␛[0m
␛[0;32mI (1151) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM␛[0m
␛[0;32mI (1158) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM␛[0m
␛[0;32mI (1164) heap_init: At 40097E78 len 00008188 (32 KiB): IRAM␛[0m
␛[0;32mI (1171) spiram: Adding pool of 4095K of external SPI memory to heap allocator␛[0m
␛[0;32mI (1179) spi_flash: detected chip: generic␛[0m
␛[0;32mI (1183) spi_flash: flash io: dio␛[0m
␛[0;32mI (1191) I2S: DMA Malloc info, datalen=blocksize=2048, dma_buf_count=16␛[0m
␛[0;32mI (1195) gpio: GPIO[36]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
␛[0;32mI (1204) gpio: GPIO[13]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
␛[0;32mI (1213) gpio: GPIO[19]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
␛[0;32mI (1223) gpio: GPIO[23]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
␛[0;32mI (1233) cpu_start: Starting scheduler on PRO CPU.␛[0m
␛[0;32mI (0) cpu_start: Starting scheduler on APP CPU.␛[0m
␛[0;32mI (1248) spiram: Reserving pool of 32K of internal memory for DMA/internal allocations␛[0m
␛[0;32I (1284) wifi:wifi driver task: 3ffd61b0, prio:23, stack:6656, core=0
I (1294) wifi:wifi firmware version: 1ba8b6a
I (1294) wifi:wifi certification version: v7.0
I (1294) wifi:config NVS flash: enabled
I (1295) wifi:config nano formating: disabled
I (1299) wifi:Init data frame dynamic rx buffer num: 32
I (1304) wifi:Init management frame dynamic rx buffer num: 32
I (1309) wifi:Init management short buffer num: 32
I (1314) wifi:Init dynamic tx buffer num: 32
I (1318) wifi:Init tx cache buffer num: 4
I (1321) wifi:Init static rx buffer size: 1600
I (1326) wifi:Init static rx buffer num: 4
I (1329) wifi:Init dynamic rx buffer num: 32
␛[0;32mI (1334) wifi_init: rx ba win: 6␛[0m
␛[0;32mI (1337) wifi_init: tcpip mbox: 32␛[0m
␛[0;32mI (1341) wifi_init: udp mbox: 16␛[0m
␛[0;32mI (1345) wifi_init: tcp mbox: 32␛[0m
␛[0;32mI (1349) wifi_init: tcp tx win: 8192␛[0m
␛[0;32mI (1353) wifi_init: tcp rx win: 32768␛[0m
␛[0;32mI (1358) wifi_init: tcp mss: 1440␛[0m
␛[0;32mI (1362) wifi_init: WiFi IRAM OP enabled␛[0m
␛[0;32mI (1366) wifi_init: WiFi RX IRAM OP enabled␛[0m
␛[0;32mI (1372) phy_init: phy_version 4771,450c73b,Aug 16 2023,11:03:10␛[0m
I (1441) wifi:mode : sta (b8:d6:1a:5a:c2:84)
I (1442) wifi:enable tsf
I (1442) wifi:Set ps type: 1

I (1451) wifi:new:<11,0>, old:<1,0>, ap:<255,255>, sta:<11,0>, prof:1
I (1452) wifi:state: init -> auth (b0)
I (1455) wifi:state: auth -> assoc (0)
I (1462) wifi:state: assoc -> run (10)
I (1471) wifi:<ba-add>idx:0 (ifx:0, e8:d1:1b:7c:35:c5), tid:0, ssn:0, winSize:64
I (1482) wifi:connected with FASTWEB-FE2QBJ-24, aid = 8, channel 11, BW20, bssid = e8:d1:1b:7c:35:c5
I (1482) wifi:security: WPA2-PSK, phy: bgn, rssi: -51
I (1488) wifi:pm start, type: 1

I (1562) wifi:AP's beacon interval = 102400 us, DTIM period = 1
␛[0;32mI (2491) esp_netif_handlers: sta ip: 192.168.1.138, mask: 255.255.255.0, gw: 192.168.1.254␛[0m
Total heap: 354675
Free heap: 212507
Total PSRAM: 4192115
Free PSRAM: 4174959-65535
ESP32 Chip model:  ESP32-D0WD-V3 Rev 3
SdkVersion:        4.4.6
␛[0;32mI (2956) I2S: I2S0, MCLK output by GPIO0␛[0m
Debug:   AudioBoard::pins::begin
Debug:   DriverPins::begin
Debug:   DriverPins::begin::setupPinMode
Debug:   pinMode 36
␛[0;32mI (2967) gpio: GPIO[36]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
Debug:   Pin 36 set
Warning: Pin '13' not set up because of conflict
Debug:   Pin 13 set
Debug:   pinMode 19
␛[0;32mI (2982) gpio: GPIO[19]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
Debug:   Pin 19 set
Debug:   pinMode 23
␛[0;32mI (3001) gpio: GPIO[23]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
Debug:   Pin 23 set
Debug:   pinMode 18
␛[0;32mI (3014) gpio: GPIO[18]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
Debug:   Pin 18 set
Debug:   pinMode 5
␛[0;32mI (3027) gpio: GPIO[5]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
Debug:   Pin 5 set
Debug:   pinMode 12
␛[0;32mI (3040) gpio: GPIO[12]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
Debug:   Pin 12 set
Debug:   pinMode 39
␛[0;32mI (3053) gpio: GPIO[39]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 ␛[0m
Debug:   Pin 39 set
Debug:   pinMode 21
␛[0;32mI (3066) gpio: GPIO[21]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 ␛[0m
Debug:   Pin 21 set
Debug:   pinMode 22
␛[0;32mI (3079) gpio: GPIO[22]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 ␛[0m
Debug:   Pin 22 set
Debug:   DriverPins::begin::SPI
Debug:   DriverPins::begin::SPI::begin
Debug:   PinsSPI::begin for 6
␛[0;32mI (3102) gpio: GPIO[13]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 ␛[0m
Info:    setting up SPI miso:2,mosi:15, clk:14, cs:13
␛[0;32mI (3113) gpio: GPIO[14]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 ␛[0m
␛[0;32mI (3116) gpio: GPIO[2]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 ␛[0m
␛[0;32mI (3125) gpio: GPIO[15]| InputEn: 1| OutputEn: 1| OpenDrain: 0| Pullup: 0| Pulldown: 0| Intr:0 ␛[0m
Debug:   DriverPins::begin::I2C
Debug:   DriverPins::begin::I2C::begin
Debug:   PinsI2C::begin for 7
Info:    setting up I2C scl: 32, sda: 33
Info:    Setting i2c clock: 100000
Debug:   AudioBoard::pins::begin::returned:true
Debug:   AudioBoard::driver::begin
Debug:   AudioDriver::begin:pins
Debug:   AudioDriver::begin:setSPI
Debug:   AudioDriver::begin:setConfig
Debug:   es8388_init
Debug:   i2c_bus_check: addr=0x10
Info:    Found ES8388
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x19 value=0x4
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x1 value=0x50
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x2 value=0x0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x35 value=0xA0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x37 value=0xD0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x39 value=0xD0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x8 value=0x0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x4 value=0xC0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x0 value=0x12
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x17 value=0x18
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x18 value=0x2
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x26 value=0x0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x27 value=0x90
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x2A value=0x90
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x2B value=0x80
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x2D value=0x0
Debug:   es8388_set_adc_dac_volume: 0.0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x1B value=0x0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x1A value=0x0
Info:    output_device: 3
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x4 value=0x3C
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x3 value=0xFF
Debug:   es8388_set_mic_gain: 25
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x9 value=0x88
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0xA value=0x5
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0xB value=0x2
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0xC value=0xD
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0xD value=0x2
Debug:   es8388_set_adc_dac_volume: 0.0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x10 value=0x0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x11 value=0x0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x3 value=0x9
Debug:   AudioDriver::begin::init succeeded
Debug:   mode->CODEC_MODE_BOTH
Debug:   es8388_ctrl_state_active: 3, 1
Debug:   es8388_start
Debug:   i2c_bus_read_bytes: addr=16 reglen=1 datalen=1 - reg=43
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x2B value=0x80
Debug:   i2c_bus_read_bytes: addr=16 reglen=1 datalen=1 - reg=43
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x3 value=0x0
Debug:   es8388_start default is mode:3
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x4 value=0x3C
Debug:   es8388_set_voice_mute
Debug:   i2c_bus_read_bytes: addr=16 reglen=1 datalen=1 - reg=25
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x19 value=0x0
Debug:   es8388_start default is mode:3
Debug:   start default is decode mode:3
Debug:   AudioDriver::begin::controlState succeeded
Debug:   es8388_config_i2s
Debug:   es8388_config_fmt
Debug:   i2c_bus_read_bytes: addr=16 reglen=1 datalen=1 - reg=12
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0xC value=0xC
Debug:   i2c_bus_read_bytes: addr=16 reglen=1 datalen=1 - reg=23
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x17 value=0x18
Debug:   es8388_set_bits_per_sample
Debug:   i2c_bus_read_bytes: addr=16 reglen=1 datalen=1 - reg=12
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0xC value=0xC
Debug:   i2c_bus_read_bytes: addr=16 reglen=1 datalen=1 - reg=23
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x17 value=0x18
Debug:   AudioDriver::begin::configInterface succeeded
Debug:   AudioDriver::begin:setPAPower
Info:    setPAPower pin 21 -> 1
Debug:   AudioDriver::begin:completed
Debug:   AudioBoard::driver::begin::returned:true
Debug:   volume 70
Debug:   es8388_set_voice_volume (HACK 1): 70
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x1A value=0x0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x1B value=0x0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x2E value=0x17
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x2F value=0x17
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x30 value=0x17
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x31 value=0x17
Debug:   AudioBoard::volume::set
Debug:   volume 50
Debug:   es8388_set_voice_volume (HACK 1): 50
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x1A value=0x0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x1B value=0x0
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x2E value=0x10
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x2F value=0x10
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x30 value=0x10
Debug:   i2c_bus_write_bytes: addr=0x10 reglen=1 datalen=1 - reg=0x31 value=0x10`

**Reply by pschatzmann** [reply]
How would you see that it is not working ?

**Reply by Marco1971Repo** [reply]
![20240305_191952_Redmi-Note-7.jpg](https://github.com/pschatzmann/arduino-audio-driver/assets/56395249/a10687d7-c357-42c6-bd75-e57a4ce48406)

The board itself work but the lcd connected to pins 21 22 doesn't.
Instead with the change:

    // add i2c codec pins: scl, sda, port, frequency
    addI2C(PinFunction::CODEC, 32, 33, -1, 100000U, Wire1);

it work as expected, consider as explained at the begin that the lib hd44780 to default use Wire, so if it doesn't work is because the board is using Wire despite the line of code you suggested to me.

**Comment by pschatzmann** [comment, marked as answer]
I have added now some update (=set) API. If you want to update just a field:
  ```
  // get current pin value 
  auto i2c_opt = AudioKitEs8388V1.getPins().getI2CPins(PinFunction::CODEC);
  if (i2c_opt){
    auto i2c = i2c_opt.value();
    i2c.p_wire = &Wire1;
    // update infrmation
    AudioKitEs8388V1.getPins().setI2C(i2c);
  }

```

Or if you want to provide all values

```
PinsI2C i2c{PinFunction::CODEC, 32, 33, -1, 100000U, Wire1};
AudioKitEs8388V1.getPins().setI2C(i2c);
```
I suggest to check the return code of the add and set functions: they return true if successful

**Reply by Marco1971Repo** [reply]
Thank you Phil!!!
It works.
